File: src/claw_mem/monitor/performance.py
# Copyright 2026 Peter Cheng
"""Performance monitor with latency histogram and hit-rate tracking (v2.19.0)."""

import time
import threading
import os
from collections import defaultdict
from typing import Dict, List


class PerformanceMonitor:
    """Tracks search latency, cache performance, and memory usage.

    Thread-safe for concurrent search recording.
    """

    def __init__(self):
        self._lock = threading.Lock()
        self._latencies: List[float] = []
        self._cache_hits = 0
        self._cache_misses = 0
        self._search_count = 0
        self._total_latency = 0.0
        self._min_latency = float("inf")
        self._max_latency = 0.0
        self._start_time = time.time()

    def record_search(self, latency_ms: float) -> None:
        with self._lock:
            self._search_count += 1
            self._latencies.append(latency_ms)
            self._total_latency += latency_ms
            if latency_ms < self._min_latency:
                self._min_latency = latency_ms
            if latency_ms > self._max_latency:
                self._max_latency = latency_ms

    def record_cache_hit(self) -> None:
        with self._lock:
            self._cache_hits += 1

    def record_cache_miss(self) -> None:
        with self._lock:
            self._cache_misses += 1

    def get_stats(self) -> Dict:
        with self._lock:
            latencies = sorted(self._latencies) if self._latencies else [0]
            n = len(latencies)

            def percentile(p):
                idx = int(n * p / 100)
                return latencies[min(idx, n - 1)]

            return {
                "search_count": self._search_count,
                "avg_latency_ms": round(self._total_latency / max(1, n), 3),
                "min_latency_ms": round(
                    self._min_latency if self._min_latency != float("inf") else 0, 3
                ),
                "max_latency_ms": round(self._max_latency, 3),
                "p50_latency_ms": round(percentile(50), 3),
                "p95_latency_ms": round(percentile(95), 3),
                "p99_latency_ms": round(percentile(99), 3),
                "cache_hits": self._cache_hits,
                "cache_misses": self._cache_misses,
                "cache_hit_rate": (
                    self._cache_hits / max(1, self._cache_hits + self._cache_misses)
                ),
                "uptime_seconds": round(time.time() - self._start_time, 1),
                "memory_mb": round(self._memory_usage_mb(), 2),
            }

    def _memory_usage_mb(self) -> float:
        try:
            import psutil

            return psutil.Process().memory_info().rss / 1024 / 1024
        except ImportError:
            return 0.0

    def reset(self) -> None:
        with self._lock:
            self._latencies.clear()
            self._cache_hits = 0
            self._cache_misses = 0
            self._search_count = 0
            self._total_latency = 0.0
            self._min_latency = float("inf")
            self._max_latency = 0.0
            self._start_time = time.time()

File: src/monitor/performance.ts
// Copyright 2026 Peter Cheng
// Licensed under Apache-2.0

/**
 * Performance monitor with latency histogram and hit-rate tracking.
 *
 * Tracks search latency, cache performance, and memory usage.
 * Uses process.memoryUsage() instead of psutil for simplicity.
 */

export interface PerformanceStats {
  search_count: number;
  avg_latency_ms: number;
  min_latency_ms: number;
  max_latency_ms: number;
  p50_latency_ms: number;
  p95_latency_ms: number;
  p99_latency_ms: number;
  cache_hits: number;
  cache_misses: number;
  cache_hit_rate: number;
  uptime_seconds: number;
  memory_mb: number;
}

export class PerformanceMonitor {
  private _latencies: number[];
  private _cacheHits: number;
  private _cacheMisses: number;
  private _searchCount: number;
  private _totalLatency: number;
  private _minLatency: number;
  private _maxLatency: number;
  private _startTime: number;

  constructor() {
    this._latencies = [];
    this._cacheHits = 0;
    this._cacheMisses = 0;
    this._searchCount = 0;
    this._totalLatency = 0.0;
    this._minLatency = Infinity;
    this._maxLatency = 0.0;
    this._startTime = Date.now();
  }

  /**
   * Record a search operation with its latency.
   *
   * @param latencyMs - Latency in milliseconds
   */
  recordSearch(latencyMs: number): void {
    this._searchCount++;
    this._latencies.push(latencyMs);
    this._totalLatency += latencyMs;
    if (latencyMs < this._minLatency) {
      this._minLatency = latencyMs;
    }
    if (latencyMs > this._maxLatency) {
      this._maxLatency = latencyMs;
    }
  }

  /**
   * Record a cache hit.
   */
  recordCacheHit(): void {
    this._cacheHits++;
  }

  /**
   * Record a cache miss.
   */
  recordCacheMiss(): void {
    this._cacheMisses++;
  }

  /**
   * Get current performance statistics.
   *
   * @returns PerformanceStats object
   */
  getStats(): PerformanceStats {
    const latencies = this._latencies.length > 0
      ? [...this._latencies].sort((a, b) => a - b)
      : [0];
    const n = latencies.length;

    const percentile = (p: number): number => {
      const idx = Math.floor((n * p) / 100);
      return latencies[Math.min(idx, n - 1)];
    };

    const totalCalls = this._cacheHits + this._cacheMisses;

    return {
      search_count: this._searchCount,
      avg_latency_ms: Math.round((this._totalLatency / Math.max(1, n)) * 1000) / 1000,
      min_latency_ms: Math.round((this._minLatency === Infinity ? 0 : this._minLatency) * 1000) / 1000,
      max_latency_ms: Math.round(this._maxLatency * 1000) / 1000,
      p50_latency_ms: Math.round(percentile(50) * 1000) / 1000,
      p95_latency_ms: Math.round(percentile(95) * 1000) / 1000,
      p99_latency_ms: Math.round(percentile(99) * 1000) / 1000,
      cache_hits: this._cacheHits,
      cache_misses: this._cacheMisses,
      cache_hit_rate: this._cacheHits / Math.max(1, totalCalls),
      uptime_seconds: Math.round((Date.now() - this._startTime) / 1000 * 10) / 10,
      memory_mb: Math.round(this._memoryUsageMb() * 100) / 100,
    };
  }

  /**
   * Get current process memory usage in MB.
   * Uses Node.js process.memoryUsage() instead of psutil.
   */
  private _memoryUsageMb(): number {
    try {
      const mem = process.memoryUsage();
      return mem.rss / 1024 / 1024;
    } catch {
      return 0.0;
    }
  }

  /**
   * Reset all collected metrics.
   */
  reset(): void {
    this._latencies = [];
    this._cacheHits = 0;
    this._cacheMisses = 0;
    this._searchCount = 0;
    this._totalLatency = 0.0;
    this._minLatency = Infinity;
    this._maxLatency = 0.0;
    this._startTime = Date.now();
  }
}
